### PR TITLE
bump openssl to 1.0.2l, fix issues #88, #102 and #116

### PR DESCRIPTION
**Pull request by Gh0u1L5**
OpenSSL just released 1.0.2l, and 1.0.2k cannot be downloaded any more.

### CODE REVIEW

**Comment by Gh0u1L5** [comment]
I also spend some time fixed the issues #88 and #102 , please review it.

File: lib/ykpiv.c [hunk 1]
```diff
@@ -154,8 +156,7 @@ ykpiv_rc ykpiv_connect(ykpiv_state *state, const char *wanted) {
     }
     rc = SCardConnect(state->context, reader_ptr, SCARD_SHARE_SHARED,
         SCARD_PROTOCOL_T1, &state->card, &active_protocol);
-    if(rc != SCARD_S_SUCCESS)
-    {
+    if(rc != SCARD_S_SUCCESS) {
```
Comment (klali) [inline lib/ykpiv.c]: Please drop this unrelated change.

File: lib/ykpiv.c [hunk 2]
```diff
@@ -202,6 +203,53 @@ ykpiv_rc ykpiv_connect(ykpiv_state *state, const char *wanted) {
...
+    apdu.st.lc = sizeof(aid);
+    memcpy(apdu.st.data, aid, sizeof(aid));
+
+    if((res = send_data(state, &apdu, data, &recv_len, &sw)) != YKPIV_OK) {
+      if(state->verbose) {
+        fprintf(stderr, "Failed communicating with card: '%s'\n", ykpiv_strerror(res));
+      }
+      return YKPIV_PCSC_ERROR;
+    } else if(sw == SW_SUCCESS) {
+      return ykpiv_verify(state, state->pin, &tries);
```
Comment (klali) [inline lib/ykpiv.c]: this might be problematic if state->pin is not set yet, probably needs a check for that.

File: lib/ykpiv.c [hunk 3]
```diff
@@ -704,6 +761,10 @@ ykpiv_rc ykpiv_verify(ykpiv_state *state, const char *pin, int *tries) {
   if((res = send_data(state, &apdu, data, &recv_len, &sw)) != YKPIV_OK) {
     return res;
   } else if(sw == SW_SUCCESS) {
+    if (state->pin == NULL) {
+      state->pin = malloc(strlen(pin) * sizeof(char));
```
Comment (klali) [inline lib/ykpiv.c]: this allocates 1 byte to few to state->pin, strlen() excludes the \0 (additionally one might want to consider having state->pin be pre-allocated like char pin[9] and check with strlen() if it's set already.

File: lib/ykpiv.c [hunk 4]
```diff
@@ -704,6 +761,10 @@ ykpiv_rc ykpiv_verify(ykpiv_state *state, const char *pin, int *tries) {
   if((res = send_data(state, &apdu, data, &recv_len, &sw)) != YKPIV_OK) {
     return res;
   } else if(sw == SW_SUCCESS) {
+    if (state->pin == NULL) {
```
Comment (klali) [inline lib/ykpiv.c]: This check is a little bit problematic in that it wont update the state->pin if ykpiv_verify is called again with a different pin, it still saves the old one in state->pin (we might also want to consider if state->pin should be updated on pin change).

File: lib/ykpiv.c [hunk 5]
```diff
@@ -202,6 +203,53 @@ ykpiv_rc ykpiv_connect(ykpiv_state *state, const char *wanted) {
   return YKPIV_GENERIC_ERROR;
 }
 
+ykpiv_rc ykpiv_reconnect(ykpiv_state *state) {
```
Comment (klali) [inline lib/ykpiv.c]: Please make this function static and not part of the ykpiv_ namespace (or if we really want it to be part of the API add it to the map file for exporting as well).

File: lib/ykpiv.c [hunk 6]
```diff
@@ -202,6 +203,53 @@ ykpiv_rc ykpiv_connect(ykpiv_state *state, const char *wanted) {
...
+  rc = SCardReconnect(state->card, SCARD_SHARE_SHARED,
+    SCARD_PROTOCOL_T1, SCARD_RESET_CARD, &active_protocol);
+  if(rc != SCARD_S_SUCCESS) {
+    if(state->verbose) {
+      fprintf(stderr, "SCardReconnect failed, rc=%08lx\n", rc);
+    }
+    return YKPIV_PCSC_ERROR;
+  }
+
+  {
```
Comment (klali) [inline lib/ykpiv.c]: Might want to consider if the select applet part should be abstracted out to a common function called by this and ykpiv_connect().

File: lib/ykpiv.c (hunk 1 above)
Comment (Gh0u1L5) [inline lib/ykpiv.c]: Hi klail, thanks for the quick reply. For this "if" line, if you search for all 23 "if"s in the whole file, you can find that there are only 2 "if"s written in two lines, the rest 21 "if"s are all written as "if (...) {". So I personally suggest that we should keep a consistent style, but if you still wish me to drop this change, feel free to tell me.

Comment (klali) [inline lib/ykpiv.c]: Yeah, I'm fine with the change itself, what I'm trying to do is keep the amount of different things in each PR to a minimum for easier review and merging.
In addition to this there is a bunch of if and else without {} at all that should be fixed, please drop this from this PR and make a new one for fixing these if you want (or I'll do that).

**Comment by klali** [comment]
Please stop adding more things to this PR, adding more things to this makes reviewing and merging it harder. Feel free to open more PRs however.

File: lib/ykpiv.c [hunk 7]
```diff
@@ -154,40 +184,15 @@ ykpiv_rc ykpiv_connect(ykpiv_state *state, const char *wanted) {
...
-        continue;
-      } else if(sw == SW_SUCCESS) {
-        return YKPIV_OK;
-      } else {
-        if(state->verbose) {
-          fprintf(stderr, "Failed selecting application: %04x\n", sw);
-        }
-      }
-    }
+    if (_select_application(state) != YKPIV_OK)
```
Comment (klali) [inline lib/ykpiv.c]: Please add {} (as previously noted we've been inconsistent about this but want them present.

File: lib/ykpiv.c [hunk 8]
```diff
@@ -202,6 +207,30 @@ ykpiv_rc ykpiv_connect(ykpiv_state *state, const char *wanted) {
...
+  }
+  rc = SCardReconnect(state->card, SCARD_SHARE_SHARED,
+    SCARD_PROTOCOL_T1, SCARD_RESET_CARD, &active_protocol);
+  if(rc != SCARD_S_SUCCESS) {
+    if(state->verbose) {
+      fprintf(stderr, "SCardReconnect failed, rc=%08lx\n", rc);
+    }
+    return YKPIV_PCSC_ERROR;
+  }
+  if ((res = _select_application(state)) != YKPIV_OK)
```
Comment (klali) [inline lib/ykpiv.c]: Same thing on these two ifs, please add {}

File: lib/ykpiv.c [hunk 9]
```diff
@@ -231,8 +260,7 @@ ykpiv_rc ykpiv_list_readers(ykpiv_state *state, char *readers, size_t *len) {
   }
 
   rc = SCardListReaders(state->context, NULL, readers, &num_readers);
-  if (rc != SCARD_S_SUCCESS)
-  {
+  if (rc != SCARD_S_SUCCESS) {
```
Comment (klali) [inline lib/ykpiv.c]: Please drop this change, we can go back over unrelated changes when we've merged this.

File: lib/ykpiv.c [hunk 10]
```diff
@@ -246,95 +274,119 @@ ykpiv_rc ykpiv_list_readers(ykpiv_state *state, char *readers, size_t *len) {
...
-  const unsigned char *in_ptr = in_data;
-  unsigned long max_out = *out_len;
-  ykpiv_rc res;
+static ykpiv_rc _begin_transaction(ykpiv_state *state) {
   long rc;
-  *out_len = 0;
+  ykpiv_rc res;
 
   rc = SCardBeginTransaction(state->card);
+  if((rc & 0xFFFFFFFF) == SCARD_W_RESET_CARD) {
```
Comment (klali) [inline lib/ykpiv.c]: is this masking necessary? we don't do it anywhere else.

File: lib/ykpiv.c [hunk 11]
```diff
@@ -246,95 +274,119 @@ ykpiv_rc ykpiv_list_readers(ykpiv_state *state, char *readers, size_t *len) {
...
+    if((res = _reconnect(state)) != YKPIV_OK) {
+      return res;
+    }
+    return _begin_transaction(state);
```
Comment (klali) [inline lib/ykpiv.c]: isn't a re-call to SCardBeginTransaction() cleaner here to make sure we only try to create the transaction twice?

File: lib/ykpiv.c [hunk 12]
```diff
@@ -246,95 +274,119 @@ ykpiv_rc ykpiv_list_readers(ykpiv_state *state, char *readers, size_t *len) {
...
+static ykpiv_rc _end_transaction(ykpiv_state *state) {
+  long rc = SCardEndTransaction(state->card, SCARD_LEAVE_CARD);
+  if(rc != SCARD_S_SUCCESS && state->verbose) {
+    fprintf(stderr, "error: Failed to end pcsc transaction, rc=%08lx\n", rc);
     return YKPIV_PCSC_ERROR;
   }
   return YKPIV_OK;
 }
 
+ykpiv_rc ykpiv_transfer_data(ykpiv_state *state, const unsigned char *templ,
```
Comment (klali) [inline lib/ykpiv.c]: Please fix the indentation of this function to match the rest of the file, essentially adding two spaces to the beginning of every line.

File: lib/ykpiv.c (hunk 10 above)
Comment (Gh0u1L5) [inline lib/ykpiv.c]: This masking is totally necessary, on the mac Sierra the size of type long seems a little bit different. The error code returned by SCardBeginTransaction isn't 0x80100068, it's 0xFFFFFFFF80100068, which caused a lot of confusing situations during my debug.

Comment (Gh0u1L5) [inline lib/ykpiv.c]: However, SCARD_S_SUCCESS should be fine, because it's just zero.

File: lib/ykpiv.c (hunk 1 above)
Comment (Gh0u1L5) [inline lib/ykpiv.c]: Ah, I see, I'll drop these changes and make another PR later.

File: lib/ykpiv.c [hunk 13]
```diff
@@ -282,14 +341,17 @@ ykpiv_rc ykpiv_transfer_data(ykpiv_state *state, const unsigned char *templ,
     memcpy(apdu.st.data, in_ptr, this_size);
     res = send_data(state, &apdu, data, &recv_len, sw);
     if(res != YKPIV_OK) {
+      _end_transaction(state);
       return res;
     } else if(*sw != SW_SUCCESS && *sw >> 8 != 0x61) {
+      _end_transaction(state);
```
Comment (klali) [inline lib/ykpiv.c]: The cases where we do _end_transaction() and then return YKPIV_OK should probably do return(_end_transaction()) no?

File: lib/ykpiv.c [hunk 14]
```diff
@@ -704,6 +761,11 @@ ykpiv_rc ykpiv_verify(ykpiv_state *state, const char *pin, int *tries) {
   if((res = send_data(state, &apdu, data, &recv_len, &sw)) != YKPIV_OK) {
     return res;
   } else if(sw == SW_SUCCESS) {
+    if (pin) {
+      free(state->pin);
+      state->pin = malloc(len * sizeof(char) + 1);
```
Comment (klali) [inline lib/ykpiv.c]: It would be nice to check that malloc() succeeds here, return YKPIV_MEMORY_ERROR, or maybe it should only allocate when the state allocates to avoid another error-case here..

File: lib/ykpiv.c [hunk 15]
```diff
@@ -767,7 +829,13 @@ static ykpiv_rc _change_pin_internal(ykpiv_state *state, int action, const char
 }
 
 ykpiv_rc ykpiv_change_pin(ykpiv_state *state, const char * current_pin, size_t current_pin_len, const char * new_pin, size_t new_pin_len, int *tries) {
-  return _change_pin_internal(state, CHREF_ACT_CHANGE_PIN, current_pin, current_pin_len, new_pin, new_pin_len, tries);
+  ykpiv_rc res = _change_pin_internal(state, CHREF_ACT_CHANGE_PIN, current_pin, current_pin_len, new_pin, new_pin_len, tries);
+  if (res == YKPIV_OK && new_pin != NULL) {
+    free(state->pin);
+    state->pin = malloc(new_pin_len * sizeof(char) + 1);
```
Comment (klali) [inline lib/ykpiv.c]: same note as for the other malloc(), check return

File: lib/ykpiv.c [hunk 16]
```diff
@@ -128,6 +130,34 @@ ykpiv_rc ykpiv_disconnect(ykpiv_state *state) {
   return YKPIV_OK;
 }
 
+static ykpiv_rc _select_application(ykpiv_state *state) {
```
Comment (klali) [inline lib/ykpiv.c]: The other static functions in the file have bare names, so please change this to select_application()

File: lib/ykpiv.c [hunk 17]
```diff
@@ -202,6 +209,32 @@ ykpiv_rc ykpiv_connect(ykpiv_state *state, const char *wanted) {
   return YKPIV_GENERIC_ERROR;
 }
 
+static ykpiv_rc _reconnect(ykpiv_state *state) {
```
Comment (klali) [inline lib/ykpiv.c]: please change to reconnect()

File: lib/ykpiv.c [hunk 18]
```diff
@@ -246,6 +279,35 @@ ykpiv_rc ykpiv_list_readers(ykpiv_state *state, char *readers, size_t *len) {
   return YKPIV_OK;
 }
 
+static ykpiv_rc _begin_transaction(ykpiv_state *state) {
```
Comment (klali) [inline lib/ykpiv.c]: please change to begin_transaction()

File: lib/ykpiv.c [hunk 19]
```diff
@@ -246,6 +279,35 @@ ykpiv_rc ykpiv_list_readers(ykpiv_state *state, char *readers, size_t *len) {
...
+  if(rc != SCARD_S_SUCCESS) {
+    if(state->verbose) {
+      fprintf(stderr, "error: Failed to begin pcsc transaction, rc=%08lx\n", rc);
+    }
+    return YKPIV_PCSC_ERROR;
+  }
+  return YKPIV_OK;
+}
+
+static ykpiv_rc _end_transaction(ykpiv_state *state) {
```
Comment (klali) [inline lib/ykpiv.c]: Please change to end_transaction()

**Comment by Gh0u1L5** [comment]
No problem, I've fixed those minor nits, plz have a review.

**Comment by klali** [comment]
I've just merged this in 37a2f778dc62dddb72c3dcac9fd7f8fec6e61467 I took the liberty of squashing and re-ordering some commits. Thank you!

**Comment by Gh0u1L5** [comment]
@klali Great, thank you.
